packages/vite/src/node/plugins/css.ts

```diff
@@ -614,10 +621,21 @@ export function cssPostPlugin(config: ResolvedConfig): Plugin {
         chunkCSS = resolveAssetUrlsInCss(chunkCSS, cssBundleName)
         // finalizeCss is called for the aggregated chunk in generateBundle
 
-        outputToExtractedCSSMap.set(
-          opts,
-          (outputToExtractedCSSMap.get(opts) || '') + chunkCSS,
-        )
+        // wait until all imports have collected their css
+        const importsPromises: Promise<void>[] = []
+        if (chunk.imports.length > 0) {
+          for (const fileName of chunk.imports) {
+            collected[fileName] = collected[fileName] ?? makeDeferred()
+            importsPromises.push(collected[fileName].promise)
+          }
+        }
+        Promise.all(importsPromises).then(() => {
+          outputToExtractedCSSMap.set(
+            opts,
+            (outputToExtractedCSSMap.get(opts) || '') + chunkCSS,
+          )
```

Comment:
IIUC this part can run after `outputToExtractedCSSMap` is used in `generateBundle` and that will output an intermediate state of CSS.
How about aggregating the chunk in `generateBundle` instead of `renderChunk`? We are calling `this.emitFile` in `generateChunk` already, so I guess it won't be a breaking change.
Also I guess it will make the implementation more simple as we won't need to use `makeDeferred`.

Reply:
Thanks for the comment! I've updated the PR.

However I still find the implementation a bit complicated but I don't know whether it can be improved.

I've considered only maintaining a map of "chunk file name" to "css" in `renderChunk` phase and traverse the bundle to collect CSS in `generateBundle` phase. However in `renderChunk` phase the `fileName`s of the chunks are not finalized yet, so we'll get a map like `{ 'index-{1}.js' => css }` and I can't determine which chunk in the final bundle it corresponds to. So I have to construct a tree of chunks even if it seems redundant.

Reply:
Thanks!

I see. I found we're working around that thing here as well.
https://github.com/vitejs/vite/blob/dd9d4c13202c6b639d43cad18e43d3b5d1a62fb2/packages/vite/src/node/plugins/css.ts#L734-L751

I came up with this solution ([stackblitz](https://stackblitz.com/edit/rollup-repro-cwg9ey?file=rollup.config.js)); injecting `nonFinalizedFilename` property in `chunk` object.
Also I created a issue on Rollup as I thought it's something that could be worth adding to rollup itself: https://github.com/rollup/rollup/issues/5080

Reply:
So I'd better wait until your PR gets merged into main and rebase on it?

Reply:
Hi @sapphi-red ! FYI, your solution has been intergrated into rollup(see https://github.com/rollup/rollup/pull/5086).

Reply:
Awesome! Thanks you! I created a PR to bump the rollup version: https://github.com/vitejs/vite/pull/14049